### PR TITLE
Add RepeatUntil action

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This repeats an action list until a condition is met. For example, it can repeat iterating a linear solve until it has converged, then proceed to the next action.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by wthrowe** [comment]
Can you add a test of the case where the loop executes zero times?  Otherwise, this looks nice.

**Comment by wthrowe** [comment]
That looks good.  Since no one else has commented, I think you can squash before a second review.

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsvu** [comment]
@nilsdeppe could do the second review for this?

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (kidder) [inline tests/Unit/Parallel/Actions/Test_Goto.cpp]: Please use  different phases to test `Goto` and `RepeatUntil`

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe 
you can squash immediately

**Review by kidder — DISMISSED** [review]

**Comment by wthrowe** [comment]
I'll save you the wait for Travis and tell you that your new test doesn't pass.  Please check such things before pushing.

**Comment by nilsvu** [comment]
@wthrowe It should pass now.

Often I push to have Travis on my own fork build and run tests, so that I can switch branches and work on something else locally. Sometimes I forget to cancel the upstream Travis build, sorry!

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (kidder) [inline tests/Unit/Parallel/Actions/Test_Goto.cpp]: can you put a counter inside this loop to demonstrate that it is entered exactly once... (if I understand what this is doing)

**Review by kidder — CHANGES_REQUESTED** [review]

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (nilsvu) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: Hm, what `RepeatUntil` does internally should not matter for this test. It should only make sure that the `Increment` action is not executed at all if the condition is already fulfilled from the start. That's being `CHECK`ed on line 110, or do you think it's not sufficient?

**Comment by wthrowe** [comment]
> Often I push to have Travis on my own fork build and run tests, so that I can switch branches and work on something else locally. Sometimes I forget to cancel the upstream Travis build, sorry!

The extra Travis build isn't what I care about.  It is a waste of reviewer time to review code that doesn't work.  If you want to use this method of testing please indicate that the code should not be reviewed.

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (kidder) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: yes, but it is not obvious that the `RepeatUntil` action is called. If I comment out line 105, the test will still pass...

Comment (nilsvu) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: But line 105 exists, so the actions must be called to set_terminate to true :)
How about I add a check that the while loop is called _at least_ once?

Comment (nilsdeppe) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: I agree with Larry.

Comment (nilsvu) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: check the fixup please

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
you can squash the fixup

**Comment by nilsvu** [comment]
@wthrowe @kidder @nilsdeppe rebased and squashed. Please take a look.

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (nilsdeppe) [inline tests/Unit/Parallel/Actions/Test_Goto.cpp]: This should be a strong equality if you actually want this check to be meaningful... As it stands it doesn't address what @kidder was worried about.

Comment (nilsvu) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: This check makes sure that the action is called. That's what I understood @kidder asked. The check below (testing that the counter has not incremented any further) makes sure the body of the `RepeatUntil` loop is not executed. I think this actually tests everything, don't you agree?

**Comment by wthrowe** [comment]
@kidder: I don't understand what your concern is.  Could you spell it out more explicitly?

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (kidder) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: @wthrowe my concern is that it is not obvious that the `next_action` call on 108 actually calls `RepeatUntil`. So if the implementation of ActionTesting is changed so that `RepeatUntil` is not called by 108, then without the additional CHECK on 115 the test will still pass even though `RepeatUntil` is not called.    And even this check doesn't really confirm that `RepeatUntil` is called.  Browsing  ActionTesting.hpp, I think what we need is a call to `force_next_action_to_be`  and to then call `next_action` instead of the code from lines 104-15

**Review by kidder — CHANGES_REQUESTED** [review]

File: tests/Unit/Parallel/Actions/Test_Goto.cpp
Comment (nilsvu) [reply tests/Unit/Parallel/Actions/Test_Goto.cpp]: @kidder please check the fixup. It implements what we discussed yesterday.

File: tests/Unit/ActionTesting.hpp
Comment (nilsdeppe) [inline tests/Unit/ActionTesting.hpp]: Could you have this be set to `number_of_actions_in_phase(next_phase) == 0` so it's consistent with `AlgorithmImpl`?

Otherwise, this and the fixup LGTM. I'm happy for you to squash (everything) once @kidder has signed off on the fixup

Comment (nilsvu) [reply tests/Unit/ActionTesting.hpp]: Ok. Do you think we should also rename `set_phase` -> `start_phase`?

Comment (kidder) [reply tests/Unit/ActionTesting.hpp]: `start_phase` could be interpreted as also executing the first action.  how about `go_to_beginning_of_phase` ?

I am happy with this and the fixup

Comment (nilsdeppe) [reply tests/Unit/ActionTesting.hpp]: I'll allow it ;)

Comment (nilsvu) [reply tests/Unit/ActionTesting.hpp]: fixed and squashed. haven't renamed `set_phase` yet since I'm not 100% sure about the name and that doesn't need to be done in this PR.

**Comment by nilsvu** [comment]
@kidder @nilsdeppe @wthrowe I think all of your review comments have been resolved. The Travis failure is #1719. Could you take a look?

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**
